crates/sui-rest-api/src/transactions/resolve.rs

```diff
@@ -0,0 +1,564 @@
+                    Some(*id)
+                }
+                _ => None,
+            })
+            .collect_vec();
+        let gas_coins = select_gas(
+            &state.reader,
+            resolved_transaction.gas_data().owner,
+            budget,
+            protocol_config.max_gas_payment_objects(),
+            &input_objects,
+        )?;
+        resolved_transaction.gas_data_mut().payment = gas_coins;
+    }
+
+    // Do one last dry-run
+    let SimulateTransactionResult {
+        input_objects,
+        output_objects,
+        events,
+        effects,
+        ..
+    } = executor
+        .simulate_transaction(resolved_transaction.clone())
+        .map_err(anyhow::Error::from)?;
+
+    let events = events.map(Into::into);
+    let effects = effects.into();
+
+    let input_objects = input_objects
+        .into_values()
+        .map(Into::into)
+        .collect::<Vec<_>>();
+    let output_objects = output_objects
+        .into_values()
+        .map(Into::into)
+        .collect::<Vec<_>>();
+    let balance_changes = derive_balance_changes(&effects, &input_objects, &output_objects);
+
+    ResolveTransactionResponse {
+        transaction: resolved_transaction.into(),
+        events,
+        effects,
+        balance_changes: Some(balance_changes),
+        input_objects: Some(input_objects),
+        output_objects: Some(output_objects),
+    }
+    .pipe(|response| match accept {
+        AcceptFormat::Json => ResponseContent::Json(response),
+        AcceptFormat::Bcs => ResponseContent::Bcs(response),
+    })
+    .pipe(Ok)
+}
+
+struct NormalizedPackage {
+    #[allow(unused)]
+    package: MovePackage,
+    normalized_modules: BTreeMap<String, normalized::Module>,
+}
+
+fn called_packages(
+    reader: &StateReader,
+    protocol_config: &ProtocolConfig,
+    unresolved_transaction: &UnresolvedTransaction,
+) -> Result<HashMap<ObjectId, NormalizedPackage>> {
+    let binary_config = sui_types::execution_config_utils::to_binary_config(protocol_config);
+    let mut packages = HashMap::new();
+
+    for move_call in unresolved_transaction
+        .ptb
+        .commands
+        .iter()
+        .filter_map(|command| {
+            if let Command::MoveCall(move_call) = command {
+                Some(move_call)
+            } else {
+                None
+            }
+        })
+    {
+        let package = reader
+            .inner()
+            .get_object(&(move_call.package.into()))?
+            .ok_or_else(|| ObjectNotFoundError::new(move_call.package))?
+            .data
+            .try_as_package()
+            .ok_or_else(|| {
+                RestError::new(
+                    axum::http::StatusCode::BAD_REQUEST,
+                    format!("object {} is not a package", move_call.package),
+                )
+            })?
+            .to_owned();
+
+        let normalized_modules = package.normalize(&binary_config).map_err(|e| {
```

Comment:
I don't think this is going to do the right thing with respect to upgraded packages. Normalization doesn't take the linkage or type origin tables into account, which means that if you have an upgraded package that introduces a new type, then that type's package ID is going to appear incorrectly if you fetch it from its normalized module.

I'll look out for where this is used to figure out the best way to get the right information.

Reply:
Since you are only using the signature information (and in particular the reference kind) I think this is fine, but I'll leave this comment as a general warning about using `NormalizedModule`.